packages/@aws-cdk/cdk/lib/core/tokens.ts

```diff
@@ -133,3 +179,128 @@ export function resolve(obj: any, prefix?: string[]): any {
+ */
+class TokenStringMap {
+    private readonly tokenMap: {[key: string]: Token};
+
+    constructor() {
+        const glob = global as any;
+        this.tokenMap = glob.__cdkTokenMap = glob.__cdkTokenMap || {};
+    }
+
+    /**
+     * Return a unique string for this Token
+     *
+     * Every call for the same Token will produce a new unique string, no
+     * attempt is made to deduplicate. Token objects should cache the
+     * value themselves, if required.
+     *
+     * If the token is an IRepresentableToken, it chooses its own representation
+     * string. This may be used to produce aesthetically pleasing and recognizable
+     * token representations for humans.
+     */
+    public register(token: Token | HintedToken): string {
+        const counter = Object.keys(this.tokenMap).length;
+        const representation = isHintedToken(token) ? token.representationHint : `TOKEN`;
+
+        const key = `${representation}.${counter}`;
+        if (new RegExp(`[^${VALID_KEY_CHARS}]`).exec(key)) {
+            throw new Error(`Invalid characters in token representation: ${key}`);
+        }
+
+        this.tokenMap[key] = token;
+
+        return `${BEGIN_TOKEN_MARKER}${key}${END_TOKEN_MARKER}`;
+    }
+
+    /**
+     * Replace any Token markers in this string with their resolved values
+     */
+    public resolveMarkers(s: string): any {
+        const spans = this.split(s);
+        const resolved = spans.map(this.resolveSpan.bind(this));
+
+        switch (resolved.length) {
+            case 0: return '';
+            case 1: return resolved[0];
+            default: return resolve(new FnConcat(...resolved));
+        }
+    }
+
+    /**
+     * Split a string on Token markers
+     */
+    private split(s: string): Span[] {
+        const re = new RegExp(`${BEGIN_TOKEN_MARKER}([${VALID_KEY_CHARS}]+)${END_TOKEN_MARKER}`, 'gu');
+        const ret: Span[] = [];
+
+        let rest = 0;
+        let m = re.exec(s);
+        while (m) {
+            if (m.index > rest) {
+                ret.push({ type: 'string', value: s.substring(rest, m.index) });
+            }
+
+            ret.push({ type: 'token', key: m[1] });
+
+            rest = re.lastIndex;
+            m = re.exec(s);
+        }
+
+        if (rest < s.length) {
+            ret.push({ type: 'string', value: s.substring(rest) });
+        }
+
+        return ret;
+    }
+
+    /**
+     * Resolve a single Span
+     */
+    private resolveSpan(span: Span): any {
+        if (span.type === 'string') {
+            return span.value;
+        }
+        if (!(span.key in this.tokenMap)) {
+            throw new Error(`Unrecognized token representation: ${span.key}`);
+        }
+
+        return resolve(this.tokenMap[span.key]);
+    }
+}
+
+interface StringSpan {
+    type: 'string';
+    value: string;
+}
+
+interface TokenSpan {
+    type: 'token';
+    key: string;
+}
+
+type Span = StringSpan | TokenSpan;
+
+const BEGIN_TOKEN_MARKER = '🔸🔹';
```

Comment:
Cute idea of using emojis, but I am not 100% sure it's sufficient (and elegant enough). I think it's important that when people look at this, they can reference it to something in the code base, search google/stack-overflow, etc, and have some intuition has to what's going on.

I would do something like `${Token[key]}`:

1. The `${}` is a signal that this is going to be substituted/expanded at some point.
2. It points to the `Token` class, which is responsible for this (maybe we can even do `cdk.Token[key]`
3. The `[]` indicate that this is a key to some map.

Reply:
I'm afraid of such simple strings in that they might naturally occur in strings where we DON'T want to replace them. So I'm looking for a string that's highly unlikely to occur (hence, the odd unicode codepoints).

Reply:
I wouldn't say `"${Token[*]}"` is a common string...